### PR TITLE
Krina Push

### PR DESCRIPTION
**Pull request by krina-shah**
## Issue ticket number and link
Closes #x 

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have added thorough tests (when necessary).
- [ ] I have added the right documentation (when needed). Product update? If yes, write one line about this update.

----
:books: Documentation preview :books:: https://sklearn-evaluation--222.org.readthedocs.build/en/222/

### CODE REVIEW

File: src/sklearn_evaluation/plot/clustering.py [hunk 1]
```diff
@@ -363,3 +418,25 @@ def silhouette_analysis_from_results(
...
+
+    # def draw_elbowcurve(self):
+       
+    #     if self.metric == "DB":
+    #         self.DaviesBouldin_analysis(self, self.X, self.clf, self.range_n_clusters, figsize=None, cmap="nipy_spectral", text_fontsize="medium", ax = self.ax)
+    #     elif self.metric == "CH":
+    #         self.calinski_harabasz_analysis(self, self.X, self.clf, self.range_n_clusters, figsize=None, cmap="nipy_spectral", text_fontsize="medium", ax = self.ax)
+    #     elif self.metric == "SIL":
+    #             self.silhouette_analysis(self,self.X,self.clf,self.range_n_clusters,self.metric,figsize=None,cmap="nipy_spectral",text_fontsize="medium",ax=self.ax)
+    #     else:
```
Comment (neelasha23) [inline src/sklearn_evaluation/plot/clustering.py:432]: Please remove the commented codes

**Comment by yafimvo** [comment]
@krina-shah can you please resolve conflicts?

**Comment by krina-shah** [comment]
Remove Comments

File: src/sklearn_evaluation/plot/clustering.py (hunk 1 above)
Comment (krina-shah) [reply src/sklearn_evaluation/plot/clustering.py]: Done

**Comment by krina-shah** [comment]
hi,
I guess I have done the needful.
I have remove the comments
Thank You

**Comment by yafimvo** [comment]
original issue : #147

**Comment by yafimvo** [comment]
@krina-shah please fix lint issues for the CI (run `black .` and then `flake8` and then push it back without the errors)

File: src/sklearn_evaluation/plot/clustering.py [hunk 2]
```diff
@@ -31,10 +31,10 @@
 import numpy as np
 
 from sklearn.base import clone
-from sklearn.metrics import silhouette_samples, silhouette_score
+from sklearn.metrics import silhouette_samples, silhouette_score, davies_bouldin_score, calinski_harabasz_score
 from sklearn.preprocessing import LabelEncoder
 from joblib import Parallel, delayed
-
+from sklearn_evaluation.plot.plot import AbstractComposedPlot, AbstractPlot
```
Comment (neelasha23) [inline src/sklearn_evaluation/plot/clustering.py:37]: Why are these imported?

File: src/sklearn_evaluation/plot/clustering.py [hunk 3]
```diff
@@ -365,3 +409,14 @@ def silhouette_analysis_from_results(
     ax.tick_params(labelsize=text_fontsize)
     ax.legend(loc="best", fontsize=text_fontsize)
     return ax
+
+
+
+
```
Comment (neelasha23) [inline src/sklearn_evaluation/plot/clustering.py:415]: Additional newlines. Please run flake8 to find these issues

File: tests/test_clustering.py [hunk 4]
```diff
@@ -303,3 +303,100 @@ def test_from_results_call(monkeypatch):
...
+
+    with pytest.raises(ValueError, match=ploomber_value_error_message):
+        plot.calinski_harabasz_analysis_from_results(n_clusters, sum_of_squares, times=None)
+
+
+def test_cluster_ranges():
+    clf = KMeans()
+    plot.calinski_harabasz_analysis(X, clf, range_n_clusters=range(1, 10))
+
+    # test old attribute doesn't break
```
Comment (neelasha23) [inline tests/test_clustering.py:357]: this is a new feature. which old attribute is being referred to here?

File: tests/test_clustering.py [hunk 5]
```diff
@@ -303,3 +303,100 @@ def test_from_results_call(monkeypatch):
...
+def test_ax_calinski_harabasz():
+    clf = KMeans()
+    fig, ax = plt.subplots(1, 1)
+    out_ax = plot.calinski_harabasz_analysis(X, clf, ax=ax)
+    assert ax is out_ax
+
+
+def test_n_jobs_calinski_harabasz():
+    clf = KMeans()
+    plot.calinski_harabasz_analysis(X, clf, n_jobs=2)
```
Comment (neelasha23) [inline tests/test_clustering.py:402]: This method doesn't seem to have a `n_jobs` parameter. Why is this test needed?

File: tests/test_clustering.py [hunk 6]
```diff
@@ -303,3 +303,100 @@ def test_from_results_call(monkeypatch):
...
+    y = np.array([0, 1, 0, 1])
+    plot.calinski_harabasz_analysis_from_results(X, y, times=None)
+
+
+@image_comparison(
+    baseline_images=["calinski_harabasz_analysis_from_results"], extensions=["png"], remove_text=False
+)
+def test_calinski_harabasz_from_results_unsorted():
+    # n_clusters = [5, 3, 9, 1, 7]
+    # sum_of_squares = np.array([389.9, 470.7, 305.5, 4572.2, 335.1])
```
Comment (neelasha23) [inline tests/test_clustering.py:387]: remove comment

File: tests/test_clustering.py [hunk 7]
```diff
@@ -303,3 +303,100 @@ def test_from_results_call(monkeypatch):
...
+
+    plot.calinski_harabasz_analysis(X, clf, range_n_clusters=range(1, 4), show_cluster_time=False)
+
+
+@image_comparison(
+    baseline_images=["calinski_harabasz_analysis_from_results"], extensions=["png"], remove_text=False
+)
+def test_calinski_harabasz_from_results():
+    # n_clusters = range(1, 10, 2)
+    # sum_of_squares = np.array([4572.2, 470.7, 389.9, 335.1, 305.5])
```
Comment (neelasha23) [inline tests/test_clustering.py:376]: remove comment

File: tests/test_clustering.py [hunk 8]
```diff
@@ -303,3 +303,100 @@ def test_from_results_call(monkeypatch):
...
+
+
+
+
+
+def test_plot_calinski_harabasz_n_clusters_future_warning():
+    clf = KMeans()
+    with pytest.warns(
+        FutureWarning,
+        match="elbow_curve will change its signature."
```
Comment (neelasha23) [inline tests/test_clustering.py:321]: Elbow curve warning shouldn't be here.

File: tests/test_clustering.py [hunk 9]
```diff
@@ -303,3 +303,100 @@ def test_from_results_call(monkeypatch):
...
+        " Please use range_n_clusters instead of n_cluster",
+    ):
+        plot.calinski_harabasz_analysis(X, clf, n_clusters=range(1, 10))
+
+
+def test_plot_calinski_harabasz_n_clusters_attribute_error():
+    clf = KMeans()
+    with pytest.raises(
+        AttributeError,
+        match="n_cluster attribute is deprecated. Please use only range_n_clusters.",
```
Comment (neelasha23) [inline tests/test_clustering.py:331]: This is a new method. Deprecated attribute not aplicable

File: src/sklearn_evaluation/plot/__init__.py [hunk 10]
```diff
@@ -30,6 +34,7 @@
 from .feature_ranking import Rank1D, Rank2D
 
 __all__ = [
+    "ElbowCurveMetrics"
```
Comment (neelasha23) [inline src/sklearn_evaluation/plot/__init__.py:37]: Why is this needed?

**Comment by krina-shah** [comment]
Hi,
I have made some changes in code.
Should I do a pull request again? or after the code review at 12?

Thank You,
Krina Shah

**Comment by yafimvo** [comment]
@krina-shah push the code so the CI will pass then we'll discuss it

**Comment by idomic** [comment]
Closing.
